### PR TITLE
[1169][ADD] Add Test-Case for stock_virtual_available_adj

### PR DESCRIPTION
**Pull request by YogeshMahera-SerpentCS**
Task #[1169](https://www.quartile.co/web#id=1169&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

File: stock_virtual_available_adj/tests/test_product.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+        # In the sales order I add some sales order lines.
+        self.sale_order_line.create(
+            {
+                "product_id": self.product_01.id,
+                "price_unit": 100.00,
+                "product_uom": self.product_uom,
+                "product_uom_qty": 20.0,
+                "order_id": self.sale_order.id,
+                "state": "sent",
```
Comment (TimLai125) [inline stock_virtual_available_adj/tests/test_product.py]: Let's make one order to be 'draft'

File: stock_virtual_available_adj/tests/test_product.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        # Call the `_compute_website_sale_available_qty`
+        # method for product 01(product.template)
+        self.product_01.product_tmpl_id._compute_website_sale_available_qty()
+
+        # Compare the `website_sale_available_qty`
+        # qty for product 01(product.template)
+        self.assertEqual(
+            self.product_01.product_tmpl_id.website_sale_available_qty,
+            140,
```
Comment (TimLai125) [inline stock_virtual_available_adj/tests/test_product.py]: Why here is 140?

File: stock_virtual_available_adj/tests/test_product.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        )
+
+        # Call the `_compute_website_sale_available_qty`
+        # method for product 02(product.template)
+        self.product_02.product_tmpl_id._compute_website_sale_available_qty()
+
+        # Compare the `website_sale_available_qty` qty for product 02(product.template)
+        self.assertEqual(
+            self.product_02.product_tmpl_id.website_sale_available_qty,
+            90,
```
Comment (TimLai125) [inline stock_virtual_available_adj/tests/test_product.py]: Why here is 90?

**Review by TimLai125 — CHANGES_REQUESTED** [review]

File: stock_virtual_available_adj/tests/test_product.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+class TestProduct(common.TransactionCase):
+    """
+        This tests allows to check website sale available quantity for product.
+    """
+
+    def setUp(self):
+        super(TestProduct, self).setUp()
+
+        self.sale_order_line = self.env["sale.order.line"]
+        self.stock_location = self.env.ref("stock.stock_location_stock")
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]:
```suggestion
        stock_location = self.env.ref("stock.stock_location_stock")
```

Avoid attaching the value to self unless it's necessary.

File: stock_virtual_available_adj/tests/test_product.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        self.product_02 = self.env["product.product"].create(
+            {"name": "Test Product 02", "type": "product"}
+        )
+        self.product_03 = self.env["product.product"].create(
+            {"name": "Test Product 03", "type": "product"}
+        )
+
+        # update the quantity on hand for products
+        self.env["stock.quant"]._update_available_quantity(
+            self.product_01, self.stock_location, 100
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]:
```suggestion
            self.product_01, stock_location, 100
```

File: stock_virtual_available_adj/tests/test_product.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        )
+        self.env["stock.quant"]._update_available_quantity(
+            self.product_02, self.stock_location, 100
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]:
```suggestion
            self.product_02, stock_location, 100
```

File: stock_virtual_available_adj/tests/test_product.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+class TestProduct(common.TransactionCase):
+    """
+        This tests allows to check website sale available quantity for product.
+    """
+
+    def setUp(self):
+        super(TestProduct, self).setUp()
+
+        self.sale_order_line = self.env["sale.order.line"]
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]:
```suggestion
```
It's pointless to declare this when it doesn't contribute to simplify/shorten the code.

File: stock_virtual_available_adj/tests/test_product.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        self.stock_location = self.env.ref("stock.stock_location_stock")
+        self.product_uom = self.env.ref("product.product_uom_unit").id
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]:
```suggestion
        self.product_uom_id = self.env.ref("product.product_uom_unit").id
```
According to https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#variable-names.

File: stock_virtual_available_adj/tests/test_product.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
+# Copyright 2020 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.tests import common
+
+
+class TestProduct(common.TransactionCase):
+    """
+        This tests allows to check website sale available quantity for product.
+    """
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]:
```suggestion
```
I think we should remove this comment as it's not carrying much meaningful message.

File: stock_virtual_available_adj/tests/test_product.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+    def test_compute_website_sale_available_qty_01(self):
+        """
+            This test compute the website
+            sale available quantity for product
+        """
+
+        # This test perform to create two sales orders which has few
+        # sale order lines and this sale order lines in `draft` and
+        # `sent` state and compute their quantity.
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]: Consolidate the comments. The bottom sentence looks broken.

File: stock_virtual_available_adj/tests/test_product.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+            }
+        )
+        self.sale_order_line.create(
+            {
+                "product_id": self.product_03.id,
+                "price_unit": 100.00,
+                "product_uom": self.product_uom,
+                "product_uom_qty": 10.0,
+                "order_id": self.sale_order.id,
+                "state": "draft",
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]: In reality you can't apply different states to lines under the same order. We should create a separate sales order as necessary.

File: stock_virtual_available_adj/tests/test_product.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+            }
+        )
+
+        # Call the `_compute_sale_quantities`
+        # method for product 03(product.product)
+        self.product_01._compute_sale_quantities()
```
Comment (yostashiro) [inline stock_virtual_available_adj/tests/test_product.py]: Why are products inconsistent here?

Besides, you should not have to call `_compute_sale_quantities` method here (I think you keep making the same mistake of calling methods unnecessarily in tests).

**Review by yostashiro — CHANGES_REQUESTED** [review]
@YogeshMahera-SerpentCS I stopped this review in the middle as the content looks broken in a few places. Please improve the quality to the point it's good for a review.

**Pull request merged**